IFC-1823 Store the display_label and HFID of node attributes in the database

gmazoyer opened the pull request:
This PR brings several changes to how `display_label` and `human_friendly_id` for nodes are stored and retrieved. While it tries to change the least amount of code to keep code compatible as much as possible, it now stores the `display_label` and `human_friendly_id` of nodes into the database to be able to retrieve them when looking up nodes instead of computing them every times. It also introduces events to help in `display_label` computation.

This PR aims to be merged into https://github.com/opsmill/infrahub/pull/7197.

This PR provides implementations for IFC-1821, IFC-1822 and IFC-1823.

gmazoyer committed:
Store the display_label of nodes in the database

gmazoyer committed:
Fix merge with display_label

ogenstad commented on backend/infrahub/core/node/__init__.py:
```diff
@@ -785,9 +789,18 @@ async def _update(
+        if self._schema.namespace != "Schema":
+            self._display_label = String(
+                name="display_label",
+                schema=TextAttributeSchema(name="display_label", kind="Text", branch=BranchSupportType.AWARE),
```
I think we might want `self._schema.branch` here instead of hardcoding `branch=BranchSupportType.AWARE)`

ogenstad commented on backend/infrahub/core/node/__init__.py:
```diff
@@ -785,9 +789,18 @@ async def _update(
         save_at = Timestamp(at)
 
+        if self._schema.namespace != "Schema":
+            self._display_label = String(
```
One thing here is that we'd run this each time we save the node even if the change isn't required. I think it makes sense to have this on node creation and then within the update it should only happen if the fields in use by the display label have been updated.

gmazoyer committed:
Fix display_label not updated after updating node

gmazoyer committed:
Use branch supprot type from schema

ogenstad committed:
Add events for display labels

gmazoyer committed:
Add `DisplayLabel` class

This class is used to keep track of the display label on a per node
basis. It registers which fields are used to build the display label to
be able to know if the display label needs to be updated when the node
is.

gmazoyer committed:
Revert a tweak

gmazoyer committed:
Create display label only on node create

gmazoyer committed:
Replicate display_label behaviour for HFID

gmazoyer committed:
Fix merge

gmazoyer committed:
Refactor some code to avoid duplicated code blocks

gmazoyer committed:
Rename variables

gmazoyer committed:
Add methods to update display_label/hfid

gmazoyer committed:
Fix var ref

gmazoyer committed:
Do not fail HFID on enum

gmazoyer committed:
Return stored HFID if it exists

gmazoyer committed:
Add `get_display_label` method like `get_hfid`

gmazoyer committed:
Extract in constants

gmazoyer committed:
Ensure classic display label is not broken

gmazoyer committed:
Ensure HFID is set for nodes

gmazoyer committed:
Revert changes to tag schema

gmazoyer committed:
Adapt diff related tests

gmazoyer committed:
Add test with new display label set

gmazoyer committed:
Fix when using enum in plain display label

gmazoyer committed:
Fix graphql default value for display label if not provided

gmazoyer committed:
Adapted more tests

gmazoyer committed:
More tweaks

gmazoyer committed:
Fix module path

ogenstad committed:
Corrections for tests

ogenstad committed:
Merge pull request #7286 from opsmill/pog-display-label-event-tests

Corrections for tests

gmazoyer committed:
Fix merge of schema nodes

gmazoyer committed:
Fix expected error messages in tests

gmazoyer committed:
More fixes for display label validation

gmazoyer committed:
One more adapted test

gmazoyer commented on backend/infrahub/core/diff/merger/serializer.py:
```diff
@@ -155,6 +156,10 @@ async def serialize_diff(
                 continue
             serial_attr_diffs = []
             for attr_diff in node.attributes:
+                if attr_diff.name in NODE_ATTRIBUTES_TO_IGNORE and not node.kind.startswith("Schema"):
+                    # We ignore attributes on actual node and not schema nodes
+                    # These ignored attributes are ignored because we rely on recomputing them after a merge operation
+                    continue
```
I wonder if we have a better way to handle this.

ajtmccarty replied:
this is probably the easiest way to do it. another option would be to exclude these kind of changes from the cypher queries to calculate a diff, but that would be significantly harder. you could also maybe use an enricher like `DiffLabelsEnricher` to remove the display labels and HFID earlier in the diff process, but it would be the same result
it would be good to add an `is_schema` method/property to the `EnrichedDiffNode` to use in this conditional

ogenstad commented on backend/infrahub/core/node/node_property_attribute.py:
```diff
@@ -0,0 +1,250 @@
+        if self.template is None:
+            return False
+
+        return bool(re.search(r"{{.+}}", self.template) or re.search(r"{%.+%}", self.template))
```
Here we'd also want some broader definition of what a Jinja2 template is to account for broken templates.

```python
return bool("{" in self.template or "}" in self.template)
```

ogenstad commented on backend/infrahub/core/node/node_property_attribute.py:
```diff
@@ -0,0 +1,250 @@
+        if items[0] in self.node_schema.attribute_names:
+            self.node_attributes.append(items[0])
+        elif items[0] in self.node_schema.relationship_names:
+            self.node_relationships.append(items[0])
```
I don't think we support having single attribute display labels be relationships.

gmazoyer committed:
Loosen "is jinja2" logic

ajtmccarty commented on backend/infrahub/core/node/__init__.py:
```diff
@@ -710,12 +720,32 @@ async def load(
             )
             self._updated_at = Timestamp(updated_at)
 
+        if self._schema.namespace != "Schema":
```
also feels like it should be a property/method

ajtmccarty commented on backend/infrahub/core/node/__init__.py:
```diff
@@ -978,6 +1021,20 @@ async def render_display_label(self, db: InfrahubDatabase | None = None) -> str:
             return repr(self)
         return display_label.strip()
 
+    async def update_human_friendly_id(self, value: list[str] | None) -> None:
```
perhaps calling these methods `set_human_friendly_id/display_label` would make it more clear that they do not save the change to the database
not totally sure about that, just my first thought

ajtmccarty commented on backend/infrahub/core/node/node_property_attribute.py:
```diff
@@ -0,0 +1,247 @@
+            return self._value
+        return {"value": self._value}
+
+    @property
+    def value(self) -> str | None:
+        if isinstance(self._value, AttributeFromDB):
+            return self._value.value
+        return self._value
+
+    def set_value(self, value: str | None) -> None:
+        """Force the value of the display label to the given one."""
+        if isinstance(self._value, AttributeFromDB):
+            self._value.value = value
+        else:
+            self._value = value
```
seems like these ones should be `abstractmethod`s too

gmazoyer replied:
That's what I wanted to do at first, but since `_value` is different between the classes, I could not figure what the typing of these should look like. Any idea?

ajtmccarty replied:
I think using `TypeVar` and `Generic` would work
we use it in a few other places
something like this, have not tested this, but I think it's about right

```python
T = TypeVar("T")



class NodePropertyAttribute(Generic[T]):
    ...

    @abstractmethod
    def set_value(self, value: T) -> None: ...

class DisplayLabel(NodePropertyAttribute[str | None]):

    ...
    def set_value(self, value: str | None) -> None: ...

class HumanFriendlyIdentifier(NodePropertyAttribute[list[str] | None]):
    ...
    def set_value(self, value: list[str] | None) -> None: ...

```

ajtmccarty commented on backend/infrahub/core/node/__init__.py:
```diff
@@ -196,6 +203,9 @@ def __init__(self, schema: NodeSchema | ProfileSchema | TemplateSchema, branch:
         self._is_protected: bool = None
         self._computed_jinja2_attributes: list[str] = []
 
+        self._display_label: DisplayLabel | None = None
+        self._human_friendly_id: HumanFriendlyIdentifier | None = None
```
would be good to have getters for these two instance variables so that you don't have to access private variables outside of this class

ajtmccarty commented on backend/infrahub/core/schema/schema_branch_display.py:
```diff
@@ -0,0 +1,123 @@
+        return self.__class__(
+            template_based_display_labels=deepcopy(self._template_based_display_labels),
+            template_relationship_triggers=deepcopy(self._template_relationship_triggers),
+        )
```
could it be a problem if that the duplicated class will be linked to the same objects on the original class? so that a change on an object in the original will affect the object on the duplicate and vice versa?

ajtmccarty commented on backend/infrahub/graphql/mutations/display_label.py:
```diff
@@ -0,0 +1,86 @@
+        existing_label = target_node._display_label.value if target_node._display_label else None
+        print(existing_label)
+
+        # TODO: Add the proper code here and send the correct event to
+        # indicate that the display_label was updated
```
I guess this is coming in a later PR?
